lib/std/Thread/Mutex.zig

```diff
@@ -1,288 +1,282 @@
+    fn tryLock(self: *Impl) bool {
+        return os.darwin.os_unfair_lock_trylock(&self.oul);
     }
 
-    /// Acquire the mutex. Will deadlock if the mutex is already
-    /// held by the calling thread.
-    pub fn lock(m: *PthreadMutex) void {
-        switch (std.c.pthread_mutex_lock(&m.pthread_mutex)) {
-            .SUCCESS => {},
-            .INVAL => unreachable,
-            .BUSY => unreachable,
-            .AGAIN => unreachable,
-            .DEADLK => unreachable,
-            .PERM => unreachable,
-            else => unreachable,
-        }
+    fn lock(self: *Impl) void {
+        os.darwin.os_unfair_lock_lock(&self.oul);
     }
 
-    pub fn unlock(m: *PthreadMutex) void {
-        switch (std.c.pthread_mutex_unlock(&m.pthread_mutex)) {
-            .SUCCESS => return,
-            .INVAL => unreachable,
-            .AGAIN => unreachable,
-            .PERM => unreachable,
-            else => unreachable,
-        }
+    fn unlock(self: *Impl) void {
+        os.darwin.os_unfair_lock_unlock(&self.oul);
     }
 };
 
-/// This has the sematics as `Mutex`, however it does not actually do any
-/// synchronization. Operations are safety-checked no-ops.
-pub const Dummy = struct {
-    locked: @TypeOf(lock_init) = lock_init,
+const FutexImpl = struct {
+    state: Atomic(u32) = Atomic(u32).init(unlocked),
 
-    const lock_init = if (std.debug.runtime_safety) false else {};
+    const unlocked = 0b00;
+    const locked = 0b01;
+    const contended = 0b11; // must contain the `locked` bit for x86 optimization below
 
-    /// Try to acquire the mutex without blocking. Returns false if
-    /// the mutex is unavailable. Otherwise returns true.
-    pub fn tryLock(m: *Dummy) bool {
-        if (std.debug.runtime_safety) {
-            if (m.locked) return false;
-            m.locked = true;
+    fn tryLock(self: *Impl) bool {
+        // Lock with compareAndSwap instead of tryCompareAndSwap to avoid reporting spurious CAS failure.
+        return self.lockFast("compareAndSwap");
+    }
+
+    fn lock(self: *Impl) void {
+        // Lock with tryCompareAndSwap instead of compareAndSwap due to being more inline-able on LL/SC archs like ARM.
+        if (!self.lockFast("tryCompareAndSwap")) {
+            self.lockSlow();
         }
-        return true;
     }
 
-    /// Acquire the mutex. Will deadlock if the mutex is already
-    /// held by the calling thread.
-    pub fn lock(m: *Dummy) void {
-        if (!m.tryLock()) {
-            @panic("deadlock detected");
+    inline fn lockFast(self: *Impl, comptime casFn: []const u8) bool {
+        // On x86, use `lock bts` instead of `lock cmpxchg` as:
+        // - they both seem to mark the cache-line as modified regardless: https://stackoverflow.com/a/63350048
+        // - `lock bts` is smaller instruction-wise which makes it better for inlining
+        if (comptime builtin.target.cpu.arch.isX86()) {
+            return self.state.bitSet(@ctz(u32, locked), .Acquire) == 0;
```

Comment:
stage2 ignores the type parameter of `@clz` in preparation for an upcoming language change. Here's how to solve:

```suggestion
            return self.state.bitSet(@ctz(u32, @as(u32, locked)), .Acquire) == 0;
```